mlshvch/hometask-4/my_hash.rb

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+require 'digest'
+
+# class MyHash with custom implementation of Hash Table
+class MyHash
+  include Enumerable
+
+  EMPTY_HASH_TABLE_SIZE = 10
+  HASH_LOAD_FACTOR = 0.7
+
+  def initialize
+    create_empty_hash
+  end
+
+  def []=(key, value)
+    rehash_table if requires_resizing?
+    index = calculate_hash_table_index(key)
+    write_to_hash(index, key, value)
+    @hash_table[index].assoc(key)[1]
+  end
+
+  def [](key)
+    index = calculate_hash_table_index(key)
+    key_value_pair = @hash_table[index].assoc(key)
+    return nil unless key_value_pair
+
+    key_value_pair.last
+  end
+
+  def delete(key)
+    index = calculate_hash_table_index(key)
+    key_value_pair = @hash_table[index].assoc(key)
+    return nil unless key_value_pair
+
+    @number_of_entries -= 1
+    @hash_table[index].delete(key_value_pair).last
+  end
+
+  def clean
+    create_empty_hash
+  end
+
+  def size
+    @number_of_entries
+  end
+
+  private
+
+  def requires_resizing?
+    (@number_of_entries.to_f / @hash_table_size) > HASH_LOAD_FACTOR
+  end
+
+  def rehash_table
+    @hash_table += Array.new(@hash_table_size, [])
+    @hash_table_size *= 2
+    @number_of_entries = 0
+    @hash_table.each_index do |index|
```

Comment:
Why not plain `each`?
```suggestion
    @hash_table.each do |bucket|
```
and use bucket instead of accessing @hash_table[index] again and again.

The biggest concern I wanted to raise is the amount of operations performed.
Your solution modifies the current hash table and then goes through the values, moving them to new places by removing from original bucket and adding to the new one. I think it could be slightly better if you created the new hash table first, copied the values over to it and then just swapped one @hash_table with another.

Let's see an example:
Let's pretend that we have a hash of 100 elements, each in their own bucket. In worst case your solution it will take around 200 * (1 deletion + 1 store) = 400 operations. (we will pretend that resizing the array and hashing is free at the moment and worst case means that rehashed values will shift by 100 elements, i.e. 1 element will become 101).
While the "recreate and copy" example will take only a 100 operations, since you iterate only through existing values and just put rehashed value into appropriate bucket.

Does it sound better, what do you think?

Reply:
So that how I did commit ago!

Reply:
Don't get frustrated, it happens sometimes. The previous mistake was to merge the "old" hash table into "new" one just to immediately rehash the values and move them somewhere else, while I suggested creating an expanded version of the hash table, copying the from old to new values with rehashing and then swapping the hash tables.
It's actually very easy.

Reply:
I don't understand: do I need to copy hash table to variable, clear the hash, and after that insert every key-value pair pair from variable back to hash?

Reply:
If by "clearing the hash" you mean "replace with expanded, but empty hash table", then yes, you are correct.